processing/src/main/java/org/apache/druid/segment/AggregateProjectionMetadata.java

```diff
@@ -0,0 +1,465 @@
+     * name will be added to {@link Projections.ProjectionMatchBuilder#remapColumn(String, String)} so the query
+     * virtual column name can be mapped to the projection physical column name. If no equivalent virtual column exists,
+     * but the inputs are available on the projection to compute it, it will be added to
+     * {@link Projections.ProjectionMatchBuilder#addReferenceedVirtualColumn(VirtualColumn)}.
+     * <p>
+     * Finally, if the column is not a virtual column in the query, it is checked with
+     * {@link Projections.PhysicalColumnChecker} which true if the column is present on the projection OR if the column
+     * is NOT present on the base table (meaning missing columns that do not exist anywhere do not disqualify a
+     * projection from being used).
+     *
+     * @param matchBuilder          match state to add mappings of query virtual columns to projection physical columns
+     *                              and query virtual columns which still must be computed from projection physical
+     *                              columns
+     * @param column                Column name to check
+     * @param queryVirtualColumns   {@link VirtualColumns} from the {@link CursorBuildSpec} required by the query
+     * @param physicalColumnChecker Helper to check if the physical column exists on a projection, or does not exist on
+     *                              the base table
+     * @return {@link Projections.ProjectionMatchBuilder} with updated state per the rules described above, or null
+     * if the column cannot be matched
+     */
+    @Nullable
+    private Projections.ProjectionMatchBuilder matchRequiredColumn(
+        Projections.ProjectionMatchBuilder matchBuilder,
+        String column,
+        VirtualColumns queryVirtualColumns,
+        Projections.PhysicalColumnChecker physicalColumnChecker
+    )
+    {
+      final VirtualColumn buildSpecVirtualColumn = queryVirtualColumns.getVirtualColumn(column);
+      if (buildSpecVirtualColumn != null) {
+        // check to see if we have an equivalent virtual column defined in the projection, if so we can
+        final VirtualColumn projectionEquivalent = virtualColumns.findEquivalent(buildSpecVirtualColumn);
+        if (projectionEquivalent != null) {
+          if (!buildSpecVirtualColumn.getOutputName().equals(projectionEquivalent.getOutputName())) {
+            matchBuilder.remapColumn(
+                buildSpecVirtualColumn.getOutputName(),
+                projectionEquivalent.getOutputName()
+            );
+          }
+          return matchBuilder;
+        }
+
+        matchBuilder.addReferenceedVirtualColumn(buildSpecVirtualColumn);
+        final List<String> requiredInputs = buildSpecVirtualColumn.requiredColumns();
+        if (requiredInputs.size() == 1 && ColumnHolder.TIME_COLUMN_NAME.equals(requiredInputs.get(0))) {
+          // wtb some sort of virtual column comparison function that can check if projection granularity time column
```

Comment:
please clean up this comment

Reply:
oops, forgot about this one since it wasn't marked with `// todo (clint): ...` like most of my ramblings